File: nachos/src/main/java/com/hootsuite/nachos/tokenizer/BaseChipTokenizer.java
package com.hootsuite.nachos.tokenizer;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.Spanned;
import android.util.Pair;

import com.hootsuite.nachos.ChipConfiguration;
import com.hootsuite.nachos.chip.Chip;

import java.util.ArrayList;
import java.util.List;

/**
 * Base implementation of the {@link ChipTokenizer} interface that performs no actions and returns default values.
 * This class allows for the easy creation of a ChipTokenizer that only implements some of the methods of the interface.
 */
public abstract class BaseChipTokenizer implements ChipTokenizer {

    @Override
    public void applyConfiguration(Editable text, ChipConfiguration chipConfiguration) {
        // Do nothing
    }

    @Override
    public int findTokenStart(CharSequence charSequence, int i) {
        // Do nothing
        return 0;
    }

    @Override
    public int findTokenEnd(CharSequence charSequence, int i) {
        // Do nothing
        return 0;
    }

    @NonNull
    @Override
    public List<Pair<Integer, Integer>> findAllTokens(CharSequence text) {
        // Do nothing
        return new ArrayList<>();
    }

    @Override
    public CharSequence terminateToken(CharSequence charSequence, @Nullable Object data) {
        // Do nothing
        return charSequence;
    }

    @Override
    public void terminateAllTokens(Editable text) {
        // Do nothing
    }

    @Override
    public int findChipStart(Chip chip, Spanned text) {
        // Do nothing
        return 0;
    }

    @Override
    public int findChipEnd(Chip chip, Spanned text) {
        // Do nothing
        return 0;
    }

    @NonNull
    @Override
    public Chip[] findAllChips(int start, int end, Spanned text) {
        return new Chip[]{};
    }

    @Override
    public void revertChipToToken(Chip chip, Editable text) {
        // Do nothing
    }

    @Override
    public void deleteChip(Chip chip, Editable text) {
        // Do nothing
    }

    @Override
    public void deleteChipAndPadding(Chip chip, Editable text) {
        // Do nothing
    }
}

File: nachos/src/main/java/com/hootsuite/nachos/tokenizer/BaseChipTokenizer.kt
package com.hootsuite.nachos.tokenizer

import android.text.Editable
import android.text.Spanned
import android.util.Pair
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.hootsuite.nachos.ChipConfiguration
import com.hootsuite.nachos.chip.Chip

/**
 * Base implementation of [ChipTokenizer] that performs no actions and returns default values.
 * Extend this class when you only need to override a subset of the interface methods.
 */
abstract class BaseChipTokenizer : ChipTokenizer {

    override fun applyConfiguration(text: Editable, chipConfiguration: ChipConfiguration) {
        /* no-op */
    }

    override fun findTokenStart(text: CharSequence, cursor: Int): Int = 0

    override fun findTokenEnd(text: CharSequence, cursor: Int): Int = 0

    @NonNull
    override fun findAllTokens(text: CharSequence): List<Pair<Int, Int>> = emptyList()

    override fun terminateToken(text: CharSequence, @Nullable data: Any?): CharSequence = text

    override fun terminateAllTokens(text: Editable) {
        /* no-op */
    }

    override fun findChipStart(chip: Chip, text: Spanned): Int = 0

    override fun findChipEnd(chip: Chip, text: Spanned): Int = 0

    @NonNull
    override fun findAllChips(start: Int, end: Int, text: Spanned): Array<out Chip> = emptyArray()

    override fun revertChipToToken(chip: Chip, text: Editable) {
        /* no-op */
    }

    override fun deleteChip(chip: Chip, text: Editable) {
        /* no-op */
    }

    override fun deleteChipAndPadding(chip: Chip, text: Editable) {
        /* no-op */
    }
} 